runtime/src/main/scala/grpcgateway/handlers/GrpcGatewayHandler.scala

```diff
@@ -6,76 +6,106 @@ import com.trueaccord.scalapb.GeneratedMessage
-              res.headers().set(HttpHeaderNames.CONTENT_TYPE, "application/json")
-              HttpUtil.setContentLength(res, json.length)
-              HttpUtil.setKeepAlive(res, HttpUtil.isKeepAlive(req))
-              ctx.writeAndFlush(res)
-
-            case Failure(e) =>
-              val status = e match {
-                case _: UnsupportedOperationException => HttpResponseStatus.NOT_FOUND
-                case _: NoSuchElementException        => HttpResponseStatus.BAD_REQUEST
-                case s: StatusRuntimeException =>
-                  s.getStatus.getCode match {
-                    case Code.OK                  => HttpResponseStatus.OK
-                    case Code.CANCELLED           => HttpResponseStatus.GONE
-                    case Code.UNKNOWN             => HttpResponseStatus.NOT_FOUND
-                    case Code.INVALID_ARGUMENT    => HttpResponseStatus.BAD_REQUEST
-                    case Code.DEADLINE_EXCEEDED   => HttpResponseStatus.GATEWAY_TIMEOUT
-                    case Code.NOT_FOUND           => HttpResponseStatus.NOT_FOUND
-                    case Code.ALREADY_EXISTS      => HttpResponseStatus.CONFLICT
-                    case Code.PERMISSION_DENIED   => HttpResponseStatus.FORBIDDEN
-                    case Code.RESOURCE_EXHAUSTED  => HttpResponseStatus.INSUFFICIENT_STORAGE
-                    case Code.FAILED_PRECONDITION => HttpResponseStatus.PRECONDITION_FAILED
-                    case Code.ABORTED             => HttpResponseStatus.GONE
-                    case Code.OUT_OF_RANGE        => HttpResponseStatus.BAD_REQUEST
-                    case Code.UNIMPLEMENTED       => HttpResponseStatus.NOT_IMPLEMENTED
-                    case Code.INTERNAL            => HttpResponseStatus.INTERNAL_SERVER_ERROR
-                    case Code.UNAVAILABLE         => HttpResponseStatus.NOT_ACCEPTABLE
-                    case Code.DATA_LOSS           => HttpResponseStatus.PARTIAL_CONTENT
-                    case Code.UNAUTHENTICATED     => HttpResponseStatus.UNAUTHORIZED
-                  }
-                case _ => HttpResponseStatus.INTERNAL_SERVER_ERROR
-              }
-              val res = new DefaultHttpResponse(req.protocolVersion(), status)
-              ctx.writeAndFlush(res).addListener(ChannelFutureListener.CLOSE)
-          }
-      } else {
-        super.channelRead(ctx, msg)
-      }
-    case _ => super.channelRead(ctx, msg)
+  def unaryCall(method: HttpMethod, uri: String, body: JValue): Future[GeneratedMessage]
+
+  override def channelRead(ctx: ChannelHandlerContext, msg: scala.Any): Unit = {
+
+    msg match {
+      case req: FullHttpRequest =>
+
+        if (supportsCall(req.method(), req.uri())) {
+
+          val f = for {
+            json <- Future {
+              val body = req.content().toString(StandardCharsets.UTF_8)
+              JsonMethods.parse(body)
+            }.recover { case err => throw new Exception(s"Wrong json payload: ${err.getMessage}") }
```

Comment:
If I'm correct this generates an `INTERNAL_SERVER_ERROR` instead of a `BAD_REQUEST`